src/metkit/api/metkit_c.cc

```diff
@@ -0,0 +1,348 @@
+#include "metkit_c.h"
+#include "metkit/mars/MarsExpension.h"
+#include "metkit/mars/MarsRequest.h"
+#include "metkit/metkit_version.h"
+
+#include "eckit/exception/Exceptions.h"
+#include "eckit/runtime/Main.h"
+#include "eckit/utils/Optional.h"
+
+#include <functional>
+
+extern "C" {
+
+// ---------------------------------------------------------------------------------------------------------------------
+
+struct metkit_marsrequest_t : public metkit::mars::MarsRequest {
+    using metkit::mars::MarsRequest::MarsRequest;
+
+    metkit_marsrequest_t(metkit::mars::MarsRequest&& k) :
+        metkit::mars::MarsRequest(std::move(k)) {}
+};
+
+struct metkit_requestiterator_t {
+    metkit_requestiterator_t(std::vector<metkit::mars::MarsRequest>&& vec) :
+        vector_(std::move(vec)), iterator_(vector_.begin()) {}
+
+    int next() {
+        if (iterator_ == vector_.end()) {
+            return METKIT_ITERATION_COMPLETE;
+        }
+        ++iterator_;
+        return iterator_ == vector_.end() ? METKIT_ITERATION_COMPLETE : METKIT_SUCCESS;
+    }
+
+    void current(metkit_marsrequest_t* request)  {
+        ASSERT(iterator_ != vector_.end());
+        *request = std::move(*iterator_);
+    }
+
+private:
+
+    std::vector<metkit::mars::MarsRequest> vector_;
+    std::vector<metkit::mars::MarsRequest>::iterator iterator_;
+};
+
+/// @comment: (maby)
+/// Not sure if there is much value in having a param iterator. We could just return an array of
+/// strings (char**) using metkit_marsrequest_params.
+/// I think we should metkit_paramiterator_t.
+struct metkit_paramiterator_t {
+    metkit_paramiterator_t(std::vector<std::string> vec) :
+        vector_(std::move(vec)), iterator_(vector_.begin()) {}
+
+    int next() {
+        if (iterator_ == vector_.end()) {
+            return METKIT_ITERATION_COMPLETE;
+        }
+        ++iterator_;
+
+        return iterator_ == vector_.end() ? METKIT_ITERATION_COMPLETE : METKIT_SUCCESS;
+    }
+
+    void current(const char** param) {
+        ASSERT(iterator_ != vector_.end());
+        *param = iterator_->c_str();
+    }
+
+private: 
+    std::vector<std::string> vector_;
+    std::vector<std::string>::iterator iterator_;
+};
+
+// ---------------------------------------------------------------------------------------------------------------------
+//                           ERROR HANDLING
+
+}  // extern "C"
+
+static thread_local std::string g_current_error_string;
+
+const char* metkit_get_error_string(int) {
+    return g_current_error_string.c_str();
+}
+
+int innerWrapFn(std::function<int()> f) {
+    return f();
+}
+
+int innerWrapFn(std::function<void()> f) {
+    f();
+    return METKIT_SUCCESS;
+}
+
+template <typename FN>
+[[nodiscard]] int tryCatch(FN&& fn) {
+    try {
+        return innerWrapFn(fn);
+    }
+    catch (const eckit::UserError& e) {
+        eckit::Log::error() << "User Error: " << e.what() << std::endl;
```

Comment:
Streams and the usage of std::endl can be problematic. In case of being unbuffered or line buffered this is fine. In case of a more sophisticated buffering strategy this can lead to additional effort.